src/sentry_sync.c

```diff
@@ -281,6 +281,77 @@ sentry__bgworker_start(sentry_bgworker_t *bgw)
     return 0;
 }
 
+typedef struct {
+    long refcount;
+    bool was_flushed;
+    sentry_cond_t signal;
+    sentry_mutex_t lock;
+} sentry_flush_task_t;
+
+static void
+sentry__flush_task(void *task_data, void *UNUSED(state))
+{
+    sentry_flush_task_t *flush_task = (sentry_flush_task_t *)task_data;
+
+    sentry__mutex_lock(&flush_task->lock);
+    flush_task->was_flushed = true;
+    sentry__cond_wake(&flush_task->signal);
+    sentry__mutex_unlock(&flush_task->lock);
+}
+
+static void
+sentry__flush_task_decref(sentry_flush_task_t *task)
+{
+    if (sentry__atomic_fetch_and_add(&task->refcount, -1) == 1) {
+        sentry__mutex_free(&task->lock);
+        sentry_free(task);
+    }
+}
+
+int
+sentry__bgworker_flush(sentry_bgworker_t *bgw, uint64_t timeout)
+{
+    if (!sentry__atomic_fetch(&bgw->running)) {
+        SENTRY_WARN("trying to flush non-running thread");
+        return 0;
+    }
+    SENTRY_TRACE("flushing background worker thread");
+
+    sentry_flush_task_t *flush_task
+        = sentry_malloc(sizeof(sentry_flush_task_t));
+    if (!flush_task) {
+        return 1;
+    }
+    flush_task->refcount = 2; // this thread + background worker
+    flush_task->was_flushed = false;
+    sentry__cond_init(&flush_task->signal);
+    sentry__mutex_init(&flush_task->lock);
+
+    sentry__mutex_lock(&flush_task->lock);
+
+    /* submit the task that triggers our condvar once it runs */
+    sentry__bgworker_submit(bgw, sentry__flush_task,
+        (void (*)(void *))sentry__flush_task_decref, flush_task);
+
+    uint64_t started = sentry__monotonic_time();
+    bool was_flushed = false;
+    while (true) {
+        was_flushed = flush_task->was_flushed;
+
+        uint64_t now = sentry__monotonic_time();
+        if (was_flushed || (now > started && now - started > timeout)) {
```

Comment:
Is the condition `now > started && now - started > timeout` written that way because of possible underflows?

Reply:
Yes, I remember we hit that even in some local debugging we did. Even though it is called monotonic time, but it is not on all platforms guaranteed to be monotonic. Similar to the workarounds rusts `Instant::now` has to jump through.